File: core/meetings/modules/recording/src/recording-codec.ts
/**
 * recording-codec — the recording.v1 MASTER codec (Node): build a final media
 * file from recording.v1 chunks. Parallels @vexa/capture-codec (the capture.v1
 * wire codec); this is the recording.v1 master codec.
 *
 * PARALLEL PATH (keep in sync). The Python twin is the pure-core module
 * `services/meeting-api/meeting_api/recording_codec.py` (NOT the finalizer — that's
 * the IO/orchestration around this). Same scope, two languages, names aligned and
 * pinned to the shared golden vectors (src/contracts/golden/):
 *      buildRecordingMaster ↔  _build_recording_master   (format dispatch)
 *      buildWebmMaster      ↔  _build_webm_master         (WebM Cluster byte-concat)
 *      buildWavMaster       ↔  _build_wav_master          (WAV RIFF header-merge)
 *      parseWavHeader       ↔  _parse_wav_header
 * Change the {webm,wav}-master logic here → mirror it there; the golden tests
 * (golden.test.ts here, test_recording_golden.py there) fail on any drift.
 * (Used by the all-Node desktop, which has no meeting-api; prod assembles via the
 * Python twin — same pattern as the desktop's node:sqlite store.)
 *
 * Two strategies, dispatched on format (the wire's recording.v1 `format`):
 *  - webm: BYTE-CONCAT in seq order. The MediaRecorder stream emits a self-
 *    describing chunk 0 (EBML + Segment + first Cluster) then Cluster-only
 *    chunks; stacking the Clusters inside the Segment yields a valid container.
 *    (ffmpeg's concat demuxer would drop the Cluster-only inputs — so NOT ffmpeg.)
 *  - wav: RIFF-aware merge — strip each chunk's 44-byte header, sum the PCM
 *    payloads, prepend one corrected master header (fmt copied from chunk 0).
 */

const WAV_HEADER_BYTES = 44;

/** Walk the RIFF chunk list (id(4) + size(4) + body, odd bodies padded to even) for `fmt ` and
 *  `data`. A standard WAV carries chunks between them — ffmpeg writes LIST/INFO by default — so
 *  `data` is located, never assumed at offset 36. `fmtChunk` is the 16-byte PCM fmt body copied
 *  verbatim into the master; a longer body (WAVE_FORMAT_EXTENSIBLE) is truncated to that prefix.
 *  The payload is the data body by its declared size, falling back to the rest of the buffer when
 *  that size is 0 or overruns — a streaming writer stamps a placeholder and appends. */
function parseWavHeader(buf: Buffer): { fmtChunk: Buffer; payload: Buffer } {
  if (buf.length < WAV_HEADER_BYTES) throw new Error("WAV chunk shorter than the 44-byte header");
  if (buf.toString("ascii", 0, 4) !== "RIFF" || buf.toString("ascii", 8, 12) !== "WAVE")
    throw new Error("WAV chunk missing RIFF/WAVE magic");

  let fmtChunk: Buffer | null = null;
  let pos = 12;
  while (pos + 8 <= buf.length) {
    const id = buf.toString("ascii", pos, pos + 4);
    const size = buf.readUInt32LE(pos + 4);
    const bodyAt = pos + 8;
    if (id === "fmt ") {
      if (bodyAt + 16 > buf.length) throw new Error("WAV 'fmt ' chunk runs past the end of the buffer");
      fmtChunk = buf.subarray(bodyAt, bodyAt + 16);
    } else if (id === "data") {
      if (fmtChunk === null) throw new Error("WAV chunk has 'data' before any 'fmt ' chunk");
      const end = size > 0 && size <= buf.length - bodyAt ? bodyAt + size : buf.length;
      return { fmtChunk, payload: buf.subarray(bodyAt, end) };
    }
    if (size > buf.length - bodyAt) break; // a size running past the buffer — nothing further parses
    pos = bodyAt + size + (size & 1);
  }
  throw new Error(`WAV chunk has no parsable 'data' chunk${fmtChunk === null ? " and no 'fmt ' chunk" : ""}`);
}

/** RIFF-aware merge (mirrors PulseAudioCapture._wrapWav). fmt is copied from the
 *  first chunk; all chunks must declare the same fmt (mismatch → throw). */
function buildWavMaster(chunks: Buffer[]): Buffer {
  const real = chunks.filter((c) => c.length >= WAV_HEADER_BYTES); // skip the empty final chunk
  if (real.length === 0) throw new Error("buildWavMaster requires at least one non-empty chunk");
  const { fmtChunk } = parseWavHeader(real[0]);
  const payloads: Buffer[] = [];
  real.forEach((c, i) => {
    const { fmtChunk: f, payload } = parseWavHeader(c);
    if (!f.equals(fmtChunk)) throw new Error(`WAV fmt chunk mismatch at chunk index ${i}`);
    payloads.push(payload);
  });
  const totalData = payloads.reduce((n, p) => n + p.length, 0);
  const header = Buffer.alloc(WAV_HEADER_BYTES);
  header.write("RIFF", 0, "ascii");
  header.writeUInt32LE(36 + totalData, 4);
  header.write("WAVE", 8, "ascii");
  header.write("fmt ", 12, "ascii");
  header.writeUInt32LE(16, 16);
  fmtChunk.copy(header, 20);
  header.write("data", 36, "ascii");
  header.writeUInt32LE(totalData, 40);
  return Buffer.concat([header, ...payloads]);
}

/** Byte-concat in seq order. The empty final chunk concatenates as a no-op. */
function buildWebmMaster(chunks: Buffer[]): Buffer {
  if (chunks.length === 0) throw new Error("buildWebmMaster requires at least one chunk");
  return Buffer.concat(chunks);
}

/**
 * Assemble `chunks` (ALREADY ordered by chunk_seq) into a single master media
 * buffer. The host writes the result to `master.<format>`.
 *
 * Note: webm output is a plain byte-concat — playable, but carries no top-level
 * duration metadata (meeting-api optionally injects it via ffmpeg; the desktop
 * keeps it dependency-free). Seeking is approximate; playback is correct.
 */
export function buildRecordingMaster(format: "webm" | "wav", chunks: Buffer[]): Buffer {
  return format === "wav" ? buildWavMaster(chunks) : buildWebmMaster(chunks);
}

File: core/meetings/services/meeting-api/src/meeting_api/recording_codec.py
"""recording.v1 MASTER codec (Python) — the PURE master-build core.

Bytes-in -> bytes-out, deterministic, no IO / DB / subprocess. This is the cloud
meeting-api's recording-assembly core: it turns the recording.v1 chunks a meeting
emitted into one master media file when the meeting ends.

PARALLEL PATH (keep in sync). This is the Python TWIN of the Node module
``meetings/modules/recording/src/recording-codec.ts`` (``buildRecordingMaster``).
Same scope, two languages, names aligned and pinned to the SHARED golden vectors
in ``meetings/modules/recording/src/contracts/golden/``::

    _build_recording_master  <->  buildRecordingMaster   (format dispatch)
    _build_webm_master       <->  buildWebmMaster        (WebM byte-concat)
    _build_wav_master        <->  buildWavMaster         (WAV RIFF header-merge)
    _parse_wav_header        <->  parseWavHeader

The deliberate two-language duplication exists because the all-Node desktop has no
Python meeting-api, while prod assembles via this twin. The goldens drift-lock the
two cores: ``test_recording_golden.py`` (here) and ``golden.test.ts`` (there) both
read the SAME vectors and must reproduce them byte-for-byte. Change a builder here
-> mirror it there, and vice-versa.

Two strategies, dispatched on the wire's recording.v1 ``format``:

* **webm** — BYTE-CONCAT in seq order. The MediaRecorder stream emits a
  self-describing chunk 0 (EBML + Segment + first Cluster) then Cluster-only
  chunks; stacking the Clusters inside the Segment yields a valid container. The
  empty final chunk concatenates as a no-op. (ffmpeg's concat demuxer would drop
  the Cluster-only inputs, so this is NOT ffmpeg.)
* **wav** — RIFF-aware merge: walk each chunk's RIFF chunk list for its ``fmt ``
  and ``data`` chunks (a standard WAV carries others between them — ffmpeg writes
  ``LIST``/``INFO`` by default), sum the ``data`` payloads, prepend one corrected
  canonical master header (``fmt`` copied from chunk 0).
"""
from __future__ import annotations

import io
import struct
from typing import List, Sequence, Tuple

# WAV / RIFF container magic: "RIFF" then size (4) then "WAVE".
_WAV_MAGIC = b"RIFF"
_WAV_FORMAT = b"WAVE"

# A canonical PCM WAV header is exactly 44 bytes:
# RIFF<sz4>WAVE fmt <16><fmt-chunk-16-bytes> data<datasz4>
_WAV_HEADER_BYTES = 44


class InvalidRecordingChunk(ValueError):
    """Uploaded bytes the codec cannot parse as the declared media format.

    A caller/data error, not a server fault: the HTTP edge maps it to 422. ``ValueError``
    subclass so anything already catching the codec's raises keeps working.
    """


def _parse_wav_header(buf: bytes) -> Tuple[bytes, bytes]:
    """Return ``(fmt_chunk_bytes, pcm_payload)`` for a WAV chunk.

    Walks the RIFF chunk list from offset 12 (id(4) + size(4) + body + a pad byte when the
    size is odd) to find ``fmt `` and ``data``. A standard WAV carries chunks between them —
    ``ffmpeg`` writes ``LIST``/``INFO`` by default — so ``data`` is located, never assumed at
    offset 36.

    ``fmt_chunk_bytes`` is the 16-byte fmt body (PCM format, channels, sample-rate, byte-rate,
    block-align, bits-per-sample) copied verbatim into the master so it inherits the source PCM
    format; a longer body (WAVE_FORMAT_EXTENSIBLE) is truncated to its 16-byte PCM prefix. The
    payload is the data body by its declared size, falling back to the rest of the buffer when the
    size is 0 or overruns — a streaming writer stamps a placeholder size and appends. Raises
    ``InvalidRecordingChunk`` for bytes that are not a parsable RIFF/WAVE with both chunks.
    """
    if len(buf) < _WAV_HEADER_BYTES:
        raise InvalidRecordingChunk(f"WAV chunk shorter than the 44-byte header: {len(buf)} bytes")
    if buf[:4] != _WAV_MAGIC or buf[8:12] != _WAV_FORMAT:
        raise InvalidRecordingChunk(f"WAV chunk missing RIFF/WAVE magic: head={buf[:12]!r}")

    fmt_body: bytes | None = None
    pos = 12
    while pos + 8 <= len(buf):
        chunk_id = buf[pos:pos + 4]
        size = struct.unpack("<I", buf[pos + 4:pos + 8])[0]
        body_at = pos + 8
        if chunk_id == b"fmt ":
            if body_at + 16 > len(buf):
                raise InvalidRecordingChunk("WAV 'fmt ' chunk runs past the end of the buffer")
            fmt_body = buf[body_at:body_at + 16]
        elif chunk_id == b"data":
            if fmt_body is None:
                raise InvalidRecordingChunk("WAV chunk has 'data' before any 'fmt ' chunk")
            end = body_at + size if 0 < size <= len(buf) - body_at else len(buf)
            return fmt_body, buf[body_at:end]
        if size > len(buf) - body_at:
            break  # a size running past the buffer — nothing further is parsable
        pos = body_at + size + (size & 1)  # RIFF pads odd-sized bodies to even
    raise InvalidRecordingChunk(
        "WAV chunk has no parsable 'data' chunk"
        + ("" if fmt_body is not None else " and no 'fmt ' chunk")
    )


def validate_wav_chunk(buf: bytes) -> None:
    """Raise ``InvalidRecordingChunk`` unless ``buf`` is a WAV chunk this codec can assemble.

    The upload edge's guard, so unparsable bytes never reach ``meeting.data``. Mirrors the
    builder's own skip rule: a chunk shorter than the 44-byte header is the bot's empty final
    chunk, which the builder drops rather than parses.
    """
    if len(buf) >= _WAV_HEADER_BYTES:
        _parse_wav_header(buf)


def _build_wav_master(chunks: Sequence[bytes]) -> bytes:
    """RIFF-aware merge (mirrors ``buildWavMaster``).

    Skips the empty final chunk, takes each remaining chunk's ``data`` payload, sums
    them, and prepends one corrected canonical master header::

        RIFF<36+total_data>WAVE fmt <16><fmt-chunk><data><total_data><payload...>

    The ``fmt`` body is copied verbatim from the FIRST non-empty chunk; every chunk
    must declare the same ``fmt`` (mismatch -> raise).
    """
    real = [c for c in chunks if len(c) >= _WAV_HEADER_BYTES]  # skip the empty final chunk
    if not real:
        raise InvalidRecordingChunk("_build_wav_master requires at least one non-empty chunk")

    fmt_chunk, _ = _parse_wav_header(real[0])
    payloads: List[bytes] = []
    for i, c in enumerate(real):
        c_fmt, payload = _parse_wav_header(c)
        if c_fmt != fmt_chunk:
            raise InvalidRecordingChunk(f"WAV fmt chunk mismatch at chunk index {i}")
        payloads.append(payload)

    total_data = sum(len(p) for p in payloads)
    out = io.BytesIO()
    out.write(_WAV_MAGIC)                          # 0..3   "RIFF"
    out.write(struct.pack("<I", 36 + total_data))  # 4..7   RIFF size = header(36) + data
    out.write(_WAV_FORMAT)                         # 8..11  "WAVE"
    out.write(b"fmt ")                             # 12..15 "fmt "
    out.write(struct.pack("<I", 16))               # 16..19 fmt chunk size = 16
    out.write(fmt_chunk)                           # 20..35 16-byte fmt body
    out.write(b"data")                             # 36..39 "data"
    out.write(struct.pack("<I", total_data))       # 40..43 data chunk size
    for p in payloads:
        out.write(p)
    return out.getvalue()


def _build_webm_master(chunks: Sequence[bytes]) -> bytes:
    """Byte-concat WebM chunks in seq order (mirrors ``buildWebmMaster``).

    The empty final chunk concatenates as a no-op.
    """
    if not chunks:
        raise InvalidRecordingChunk("_build_webm_master requires at least one chunk")
    return b"".join(chunks)


def _build_recording_master(media_format: str, chunks: Sequence[bytes]) -> bytes:
    """Dispatch to the master builder by ``format`` — the twin of
    ``buildRecordingMaster``. ``wav`` -> RIFF header-merge; anything else
    (i.e. ``webm``) -> byte-concat. Pure: ALREADY-ordered chunks -> master bytes.
    """
    if (media_format or "").lower() == "wav":
        return _build_wav_master(chunks)
    return _build_webm_master(chunks)


def build_recording_master(chunks: Sequence[bytes], media_format: str) -> bytes:
    """Public front door: assemble ALREADY-ordered recording.v1 ``chunks`` into a
    single master media buffer for ``media_format`` (``"webm"`` | ``"wav"``).

    The host writes the result to ``master.<format>``. WebM output is a plain
    byte-concat (playable; no top-level duration metadata — meeting-api optionally
    injects it via ffmpeg downstream). WAV output is a RIFF header-merge with the
    data size corrected to the summed PCM length.
    """
    return _build_recording_master(media_format, chunks)
